### PR TITLE
Added ResNet-{18, 34, 50, 101, 152}

### PR DESCRIPTION
**Pull request by guglielmocamporese**
Simple usage:

```python
from models.resnet import ResNet50

net = ResNet50(num_classes=10, pretrained=True)
x = Tensor.uniform(1, 3, 224, 224, requires_grad=False)  
y = net(x) # or net.forward(x)
# y shape is [1, 10]
```

### Main Changes:
* ResNets are defined in `models/resnet.py`,
  * `pretrained` nets are available, number of weights are checked (you can check them also on testing),
* resnets tests are implemented in `test/test_train.py`,
* Implemented some useful layes in `tinygrad.nn` like: `nn.Linear`, `nn.Conv2d`, `nn.Sequential`, `nn.Identity`, ...,
* Tests for layers in `tinygrad.nn` are implemented in `test/test_nn.py`,
* Implemented train example of a resnet in `extra/train_resnet.py`.

### Run training with ResNets:

```console
PYTHONPATH="." GPU="1" python3 examples/train_resnet.py
```

### CODE REVIEW

**Comment by geohot** [comment]
Cool!

File: extra/training.py [hunk 1]
```diff
@@ -14,14 +14,20 @@ def sparse_categorical_crossentropy(out, Y):
...
+          transform=None, target_transform=None):
   Tensor.training = True
   losses, accuracies = [], []
   for i in (t := trange(steps, disable=os.getenv('CI') is not None)):
     samp = np.random.randint(0, X_train.shape[0], size=(BS))
 
-    x = Tensor(X_train[samp])
+    x = X_train[samp]
+    if transform is not None:
+        x = transform(x)
```
Comment (geohot) [inline extra/training.py:26]: Spacing

File: tinygrad/nn.py [hunk 2]
```diff
@@ -30,3 +31,59 @@ def normalize(self, x, mean, var):
...
+  def __call__(self, x):
+    B, *dims, D = x.shape
+    x = x.reshape(shape=(B * np.prod(dims).astype(np.int32), D))
+    x = x.dot(self.weight)
+    if self.use_bias:
+      x = x.add(self.bias.reshape(shape=[1, -1]))
+    x = x.reshape(shape=(B, *dims, -1))
+    return x
+
+class Dropout:
```
Comment (geohot) [inline tinygrad/nn.py:52]: Hmm, I understand these things match pytorch...I guess there's not much harm in them

File: tinygrad/nn.py [hunk 3]
```diff
@@ -30,3 +31,59 @@ def normalize(self, x, mean, var):
...
+    self.in_dim = in_dim
+    self.out_dim = out_dim
+    self.use_bias = bias
+    self.weight = Tensor.uniform(in_dim, out_dim)
+    if self.use_bias:
+      self.bias = Tensor.zeros(out_dim)
+
+  def __call__(self, x):
+    B, *dims, D = x.shape
+    x = x.reshape(shape=(B * np.prod(dims).astype(np.int32), D))
```
Comment (geohot) [inline tinygrad/nn.py:45]: I don't think you need to reshape here, dot should broadcast across the top dimensions.

File: extra/training.py [hunk 4]
```diff
@@ -14,14 +14,20 @@ def sparse_categorical_crossentropy(out, Y):
   y = Tensor(y)
   return out.mul(y).mean()
 
-def train(model, X_train, Y_train, optim, steps, BS=128, lossfn=sparse_categorical_crossentropy):
+def train(model, X_train, Y_train, optim, steps, BS=128, lossfn=sparse_categorical_crossentropy,
```
Comment (geohot) [inline extra/training.py:17]: Instead of None, just make the default function the identity.

File: examples/train_resnet.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    for t in self.trans:
+      x = t(x)
+    return x
+
+if __name__ == "__main__":
+  model = ResNet18(num_classes=10, pretrained=True)
+
+  X_train, Y_train, X_test, Y_test = fetch_mnist()
+  lr = 0.003
+  transform = ComposeTransforms([
```
Comment (geohot) [inline examples/train_resnet.py:30]: Perhaps just preapply this? It lets you remove the complexity in the trainer

Clever idea about big mnist btw.

File: tinygrad/nn.py (hunk 3 above)
Comment (guglielmocamporese) [reply tinygrad/nn.py:45]: Yes, but I got an error without reshaping the input along the first dimension in the case the input has a rank > 2.

**Comment by geohot** [comment]
Sweet, running tests now, will merge if pass

File: test/test_train.py [hunk 6]
```diff
@@ -31,16 +32,20 @@ def test_efficientnet(self):
     train_one_step(model,X,Y)
 
   def test_transformer(self):
+    pass
```
Comment (geohot) [inline test/test_train.py:35]: Why is this commented out?

**Pull request merged**